src/renderer.ts

```diff
@@ -518,6 +529,8 @@ class Renderer extends EventEmitter<RendererEvents> {
         tailDelay(() => {
           renderTail(fromIndex + viewportLen, toIndex + viewportLen)
         })
+      } else {
+        complete('tail')
```

Comment:
I'm thinking, maybe this could be refactored as a promise, smth like:

```js

    // Draw the waveform in viewport chunks, each with a delay
    return Promise.all([
      new Promise((resolve) => {
        const headDelay = this.createDelay()
        const renderHead = (fromIndex: number, toIndex: number) => {
          draw(fromIndex, toIndex)
          if (fromIndex > 0) {
            headDelay(() => {
              renderHead(fromIndex - viewportLen, toIndex - viewportLen)
            })
          } else {
            resolve(true)
          }
        }
        renderHead(start, end)
      }),

      new Promise((resolve) => {
        if (end >= len) {
          resolve(true)
          return
        }
        const tailDelay = this.createDelay()
        const renderTail = (fromIndex: number, toIndex: number) => {
          draw(fromIndex, toIndex)
          if (toIndex < len) {
            tailDelay(() => {
              renderTail(fromIndex + viewportLen, toIndex + viewportLen)
            })
          } else {
            resolve(true)
          }
        }
        renderTail(end, end + viewportLen)
      }),
    ])
```

Reply:
@katspaugh, we thought about that, too; however, one downside to this is that createDelay could end up canceling the setTimeout, which would mean that `resolve` never gets called and ends up adding unresolved promises as memory leaks.

Reply:
@katspaugh We thought about that, too; however, one downside to this is that `createDelay` could end up canceling the setTimeout, which would mean that resolve never gets called and ends up adding unresolved promises as memory leaks.

This is especially problematic if the rendercomplete gets interrupted by a new render.

Reply:
True, good point. I’ll approve the PR as is then.